2016/proposals/narayan-aditya-visualizer.md

```diff
@@ -0,0 +1,244 @@
+ * Backend  : [Django Rest Framework]
+ * Scheduler: [Cron]
+ * Database : [PostgreSQL]
+ * Frontend : [Charts.js] and [Bootstrap]
+
+### Backend - Django Application
+
+The survey visualizer application would be made using
+[Django Rest Framework] (henceforth called DRF) over [Django]
+to allow easier integration with [AMY] when required.
+
+The tentative API structure exposed by DRF is:
+ * GET `/api/workshop/`  
+   Lists all workshops that have surveys collecting responses.
+ * GET `/api/workshop/<workshop_id>/`  
+   Returns details of a particular workshop.
+ * GET `/api/workshop/<workshop_id>/survey/`  
+   Lists all the surveys whose responses are being collected.
+   Each survey has its corresponding `survey_id`.
+ * POST `/api/workshop/<workshop_id>/survey/refetch`  
+   Repopulates the list of surveys whose response are being
+   collected.
+ * GET `/api/workshop/<workshop_id>/survey/<survey_id>`  
+   Returns the responses for the particular survey, listed
+   according to the questions contained in the survey.
+ * POST `/api/workshop/<workshop_id>/survey/<survey_id>/refetch`  
+   Instructs the Django application to fetch the responses
+   that the survey has received since the last time the
+   SurveyMonkey API was hit.
+
+The underlying Django app will perform these tasks:
+ * Polling SurveyMonkey API management command:
+   This management command would populate the database with survey
+   responses fetched from the SurveyMonkey API.
+   The number of survey responses obtained from the [get_response_counts]
+   API call will be compared to the number of survey responses
+   saved in the database. The remaining responses will be fetched
+   using the [get_respondent_list] and [get_responses] API calls.
+ * Segregating the survey responses according to the answer to the
+   question: *Which workshop did you attend?* or the custom variable
+   `workshop_id`.
+ * Rendering and serving the templates.
+
+### Scheduler - Cron
+
+[Cron] will be used as a job scheduler to periodically run
+the *Polling SM API* management command at fixed time intervals.
+The interval duration would be dynamically determined by the
+number of API calls left and the time of the day.
+
+*Why Cron?*
+ * Cron is the easiest job scheduling solution available to us
+   and is available across all Unix environments.
+ * > Premature optimization is the root of all evil
+
+   -- DonaldKnuth
+
+### Database - PostgreSQL
+
+The Django app will use [PostgreSQL] as its database engine to
+store the survey responses fetched by the backend.
+
+*Why PostgreSQL?*
+ *  Django provides [PostgreSQL-specific features] like the ability
+    and query JSONs, dicts and arrays. The database schema, when
+    finalized, would benefit from these features.
+ *  PostgreSQL is actively maintained and is seen to perform
+    faster writes as compared to SQLite.
+ *  Although AMY currently uses SQLite database, a [migration]
+    is currently being planned to upgrade AMY to PostgreSQL.
```

Comment:
I have attached a rudimentary schema design of the database.
![my_project_visualized](https://cloud.githubusercontent.com/assets/6830800/13649943/3a83d18c-e666-11e5-81f7-78ea74b26ee1.png)
The tables are explained:
 *  **Workshop**: Records details about the workshop. Admins add a row when they add a answer to the question *Which workshop did you attend?* or create a new instance of the custom variable `workshop_id`.
 *  **Survey**: Rows correspond to the number of surveys in the SM account. Fields are auto-populated using the [get_survey_details](https://developer.surveymonkey.com/docs/methods/get_survey_details/) API endpoint.
 *  **Survey_Workshop**: Acts as a through model between a survey and a workshop.  
    ```python
    class Survey(models.Model):  
        ...  
        workshops = models.ManyToManyField(  
          Workshop,  
          through=Survey_Workshop,  
          through_fields=('survey', 'workshop'),  
        )  
    ```
 *  **Response**: Lists all survey responses across all workshops in a survey.

There exists different types of questions with different schema (single choice, multiple choice, matrix choice etc.). Each question (and its corresponding answers) will have a different table with the required schema.
Relationships that point to the questions and answers will be handles using [generic relations](https://docs.djangoproject.com/en/1.9/ref/contrib/contenttypes/#generic-relations).

PROS:
* Pure SQL: No JSONs are saved. Compatible across all SQL databases.

CONS:
* I have apprehensions about storing all survey responses within one table. This table will end up containing thousands(or millions) of rows. I cannot anticipate the time taken to query rows with `survey__id = x` filter.